crates/ruff_linter/src/suppression.rs

```diff
@@ -0,0 +1,905 @@
+    pub(crate) fn new(source: &'a str) -> Self {
+        Self {
+            source,
+            ..Default::default()
+        }
+    }
+
+    fn match_comments(&mut self, current_indent: &String) {
+        let mut comment_index = 0;
+
+        // for each pending comment, search for matching comments at the same indentation level,
+        // generate range suppressions for any matches, and then discard and unmatched comments
+        // from the outgoing indentation block
+        while comment_index < self.pending.len() {
+            let comment = &self.pending[comment_index];
+
+            // skip comments from an outer indentation level
+            if comment
+                .indent
+                .as_ref()
+                .is_some_and(|indent| indent.text_len() < current_indent.text_len())
+            {
+                comment_index += 1;
+                continue;
+            }
+
+            // find the first matching comment
+            if let Some(other_index) = self.pending[comment_index + 1..]
+                .iter()
+                .position(|other| comment.matches(other, self.source))
+            {
+                // offset from current candidate
+                let other_index = comment_index + 1 + other_index;
+                let other = &self.pending[other_index];
+
+                // record a combined range suppression from the matching comments
+                let combined_range = TextRange::new(comment.range.start(), other.range.end());
+                for code in comment.codes_as_str(self.source) {
+                    self.valid.push(Suppression {
+                        code,
+                        range: combined_range,
+                        comments: smallvec![comment.clone(), other.clone()],
+                    });
+                }
+
+                // remove both comments from further consideration
+                self.pending.remove(other_index);
+                self.pending.remove(comment_index);
+            } else {
+                self.invalid.push(InvalidSuppression {
+                    kind: InvalidSuppressionKind::Unmatched,
+                    comment: self.pending.remove(comment_index).clone(),
+                });
+            }
+        }
+    }
+
+    pub(crate) fn load_from_tokens(&mut self, tokens: &Tokens) -> Suppressions {
+        let default_indent = String::new();
+        let mut current_indent: &String = &default_indent;
+        let mut indents: Vec<String> = vec![];
+
+        // Iterate through tokens, tracking indentation, filtering trailing comments, and then
+        // looking for matching comments from the previous block when reaching a dedent token.
+        for (token_index, token) in tokens.iter().enumerate() {
+            match token.kind() {
+                TokenKind::Indent => {
+                    indents.push(self.source.slice(token).to_string());
+                    current_indent = indents.last().unwrap_or(&default_indent);
```

Comment:
Now that we use `&str`, which is `Copy`, we can write this as 

```suggestion
										current_indent = self.source.slice(token);
                    indents.push(current_indent);
```

Although I'd suggest removing `current_indent` altogether and instead use `indents.last().copied().unwrap_or_default()` to get the current indent. It avoids tracking the same state twice